### PR TITLE
fix same file documentation bug

### LINKED ISSUE #1261: Documentation missing for references to things defined in the same file
Given an input, `a.ml`:

```ocaml
(** f docs *)
let f = ()

(** g docs *)
let g = ()

let h = g
```

When the point is on the reference to `g`, all is well:

```
$ ocamlmerlin single document -verbosity 0 -position 7:9 < a.ml | jq .value
"g docs"
```

However, if we change `h` to be `let h = f`, the documentation for `f` is not returned:

```
$ ocamlmerlin single document -verbosity 0 -position 7:9 < a.ml | jq .value
"No documentation available"
```

If we remove the documentation comment for `g` though (note the changed line number to correspond), the documentation for `f` _is_ found:

```
$ ocamlmerlin single document -verbosity 1 -position 6:9 < a.ml | jq .value
"f docs"
```

It seems that, for lookups of things defined in the same file, merlin will not return documentation for any but the final documented definition.

(Originally reported in https://github.com/ocamllabs/vscode-ocaml-platform/issues/519, where @ulugbekna suggested the actual failure pattern.)

### PR DESCRIPTION
**Pull request by ulugbekna**
Hi! 

Fixes #1261 

tl;dr `Ocamldoc.associate_comment` needs a list of doc-comments from the start to the end of file (non-decreasing positions of comments), but `Locate.get_doc` returned the comments from the end to the start of the file (non-increasing positions of comments). The last doc-comment was available because the last elem of non-decr list = head of non-incr list.

More detailed: 

`Ocamldoc.associate_comment` takes a list of doc-comments (with their positions) from a file and compares positions of those comments with the position `p` of the cursor. If a doc-comment `d` with position `p1` has `p1 > p` (ie documentation occurs later in the file than `p`), it considered there the list doesn't contain documentation for position `p`. (see [code](https://github.com/ocaml/merlin/blob/2a0dd5c16178efcc4b8132e3c3b3c2a6cc7b13ea/src/analysis/ocamldoc.ml#L42-L46))

Since `Locate.get_doc` was returning `List.rev comments`, it returned doc-comments in a non-decreasing order, which meant that head of the list would have position "larger" than `p` causing the `associate_comment` return `None`.

### CODE REVIEW

**Comment by voodoos** [comment]
Thanks, these changes look good to me.
I wonder why these `List.rev` was used in the first place, @trefis ?

It might just  be a mistake; or the comments list was previously in the wrong order ? Or `Ocamldoc.associate_comment` was too aggressively optimized at some point...

I am also curious about the role of `File_switching.where_am_i ()` ?

**Comment by ulugbekna** [comment]
Thanks Ulysse!

In retrospective, I should've probably added tests to show that there are
no regressions. I have tested manually however if that counts :p

**Comment by voodoos** [comment]
It's never to late to add tests 🙂

**Comment by voodoos** [comment]
This release should fix #1288, #1300, #714, #1261, #1191, ocaml/dune#4479.
